packages/gatsby/src/query/graphql-runner.js

```diff
@@ -18,23 +20,96 @@ class GraphQLRunner {
     })
   }
 
-  query(query, context) {
+  async query(query, context, { parentSpan } = {}) {
     const { schema, schemaCustomization } = this.store.getState()
 
-    return graphql(
+    const activity = new SpanTracker(`GraphQL Query`, {
+      parentSpan: parentSpan,
+      tags: {
+        queryName: query.name,
+      },
+    })
+
+    activity.start()
+    const result = await graphql({
       schema,
-      query,
-      context,
-      withResolverContext({
+      source: query.text,
+      rootValue: context,
+      contextValue: withResolverContext({
         schema,
         schemaComposer: schemaCustomization.composer,
-        context,
+        context: {
+          ...context,
+          spanTracker: activity,
+        },
         customContext: schemaCustomization.context,
         nodeModel: this.nodeModel,
       }),
-      context
-    )
+      variableValues: context,
+      fieldResolver: tracingResolver(),
+    })
+    activity.done()
+    return result
+  }
+}
+
+class SpanTracker {
+  constructor(name, { parentSpan, ...args }) {
+    this.parentSpan = tracer.startSpan(name, {
+      childOf: parentSpan,
+      ...args,
+    })
+    this.spans = {}
+  }
+
+  start() {}
+
+  done() {
+    this.parentSpan.finish()
+  }
+
+  createActivity(path, name, args) {
+    let prev = path.prev
+    if (prev && Number.isInteger(prev.key)) {
+      prev = prev.prev
+    }
+    const parentSpan = this.getSpan(prev)
+    const span = tracer.startSpan(name, { childOf: parentSpan, ...args })
+    this.setSpan(path, span)
+    return {
+      span,
+
+      start() {},
+
+      done() {
+        span.finish()
+      },
+    }
+  }
+
+  getSpan(gqlPath) {
+    const path = pathToArray(gqlPath)
```

Comment:
It's hard to understand where you do the null check but basically pathToArray is guarding us because of the while loop. Maybe make it explicit.

```js
if (!gqlPath) {
  return this.parentSpan;
}
```